### ISSUE TITLE
Broken pipe error when attempting to create a bunch of clients in one process

### ISSUE THREAD

**Issue by nicholasbishop** [hyperium/tonic#209]
tl;dr: when I create about 500 client connections in a process I get an error: "Client: connection error: broken pipe".

I realize this is a somewhat unusual thing to do, since normally you'd just create a connection and reuse it rather than creating a bunch of them. I got into this situation because I was creating a small benchmark to compare grpc-rs and tonic where I wanted to simulate a bunch of clients. I noticed that I could create a whole bunch of simultaneous connections to the grpc-rs server, but tonic would fail. (It seems to work fine if the client connections are each in their own process, this error only occurs when the connections are created from a single process.)

I've put the implementation in this repo: https://github.com/nicholasbishop/tonic-limit-bug

The core of it is pretty simple:

```rust
async fn send_request() -> Result<(), BoxError> {
    let mut client = MockClientTonic::connect("http://[::1]:50051").await?;
    let request = tonic::Request::new(DoItRequest {});
    let _reply = client.do_it(request).await?.into_inner();
    Ok(())
}

#[tokio::main]
async fn main() -> Result<(), BoxError> {
    let num_clients = get_num_clients();

    let _server = run_server()?;
    // Give the server time to start
    delay_for(Duration::from_millis(100)).await;

    let r =
        join_all((0..num_clients).map(|_| tokio::spawn(send_request()))).await;

    for r in r {
        r??;
    }

    Ok(())
}
```

**Comment by LucioFranco** [comment]
@nicholasbishop hey! Can you run this same test to reproduce it but this time add `env_logger` and run it it with `RUST_LOG=hyper=debug` and one run with `RUST_LOG=h2=debug`?

**Comment by nicholasbishop** [comment]
Sure, here are the logs:
- https://raw.githubusercontent.com/nicholasbishop/tonic-limit-bug/master/logs/hyper-debug.log
- https://raw.githubusercontent.com/nicholasbishop/tonic-limit-bug/master/logs/h2-debug.log

**Comment by LucioFranco** [comment]
Ok looks like the go server is resetting the connections.

`[2019-12-26T19:25:27Z DEBUG hyper::proto::h2::client] connection error: Connection reset by peer (os error 104)`

Your join all will fail all clients if one of them fails. So it may make sense to modify how you do the join_all and handle the errors in the spawned task.

**Comment by nicholasbishop** [comment]
Makes sense. I am surprised that I don't get the same behavior when launching a bunch of clients in separate processes, but perhaps the extra time spent launching the processes is enough to avoid hammering the server with quite as many simultaneous connections.
